README.md
mumbl
=====

*Version 0.1.1b1*

**mumbl** is a JavaScript library that makes it easy to play music and create playlists
on web pages.


Demo
----

A demo is included with mumbl but if you dont want to download it, there is also an
[online demo][demo].

Please note that mumbl is not the player in the demo. mumbl is the *back-end* and the
demo is just an example of using mumbl.


Supported Platforms
-------------------

* HTML5 &lt;audio/&gt;
* [Songbird][sb]
* [SoundManager 2][sm2] version 2.95b.20100323 or higher


Supported Browsers
------------------

* HTML5
    * Firefox 3.5+
    * Google Chrome 4+
* SoundManager 2 (version 2.95b.20100323+)
    * Firefox 1.5+
    * Opera 10+
    * Google Chrome 1+
* Songbird 1.4+


API
---

Strong and emphasized text has titles (which can be viewed by hovering your cursor over
them) containing their type if they are not functions or return type if they are.


### Methods

<dl>
  <dt><code>mumbl.<strong title="void">addTrack</strong>(<strong title="Array.&lt;?String&gt;">TrackItem</strong>)</code></dt>
  <dd>Adds <code>TrackItem</code>* to the playlist.</dd>

  <dt><code>mumbl.<strong title="void">addTracks</strong>(<strong title="Array.&lt;TrackItem&gt;">TrackItemList</strong>)</code></dt>
  <dd>
    Adds every <code>TrackItem</code> in <code>TrackItemList</code> to the playlist.
  </dd>

  <dt><code>mumbl.<strong title="void">clear</strong>()</code></dt>
  <dd>
    Clears the playlist.
  </dd>

  <dt><code>mumbl.<strong title="Function">createEventDispatcher</strong>(<strong title="String">event</strong>)</code></dt>
  <dd>
    Creates an event dispatcher for the specified event. Custom events cannot be observed
    until at least one event dispatcher for the event has been created.
  </dd>

  <dt><code>mumbl.<strong title="void">destruct</strong>()</code></dt>
  <dd>
    Uninitializes mumbl. <code>window.mumbl</code> will be deleted.
  </dd>

  <dt><code>mumbl.<strong title="Number (non-negative)">duration</strong>()</code></dt>
  <dd>
    Returns the duration in seconds of the currently selected track.
  </dd>

  <dt><code>mumbl.<strong title="Number (non-negative integer)">length</strong>()</code></dt>
  <dd>
    Returns the amount of tracks in the playlist.
  </dd>

  <dt><code>mumbl.<strong title="Boolean">mute</strong>()</code></dt>
  <dd>
    Returns <code>true</code> if the player is muted or <code>false</code> if the
    player is not muted.
  </dd>

  <dt><code>mumbl.<strong title="void">mute</strong>(<strong title="Boolean">muteState</strong>)</code></dt>
  <dd>
    If <code>muteState</code> is <code>true</code>, the player is muted if not
    already muted.
    <br />
    If <code>muteState</code> is <code>false</code>, the player is unmuted if not
    already unmuted.
  </dd>

  <dt><code>mumbl.<strong title="void">observe</strong>(<strong title="String">event</strong>, <strong title="Function">observer</strong>)</code></dt>
  <dd>
    This calls <code>observer</code>, passing it the event type as the first argument,
    whenever the specified event is dispatched.
   </dd>

  <dt><code>mumbl.<strong title="Number (integer from 0 to 2)">loop</strong>()</code></dt>
  <dd>
    Returns the current <code>loopType</code>.
  </dd>

  <dt><code>mumbl.<strong title="void">loop</strong>(<strong title="Number (integer from 0 to 2)">loopType</strong>)</code></dt>
  <dd>
    If <code>loopType</code> is <code>0</code>, looping is turned off.
    <br />
    If <code>loopType</code> is <code>1</code>, the currently selected track is looped.
    <br />
    If <code>loopType</code> is <code>2</code>, the playlist is looped. This is the
    default <code>loopType</code>.
  </dd>

  <dt><code>mumbl.<strong title="void">next</strong>()</code></dt>
  <dd>Selects the next track in the playlist from the currently selected track.</dd>

  <dt><code>mumbl.<strong title="void">onready</strong>(<strong title="Function">callback</strong> [, <strong title="?Object">scope</strong>])</code></dt>
  <dd>
    After mumbl is ready to be used, the callback is called using the scope specified or
    the <code>mumbl</code> object if no scope is specified.
  </dd>

  <dt><code>mumbl.<strong title="void">play</strong>()</code></dt>
  <dd>
    Plays the currently selected track at its current position.
  </dd>

  <dt><code>mumbl.<strong title="void">pause</strong>()</code></dt>
  <dd>
    Pauses playback of the currently selected track.
  </dd>

  <dt><code>mumbl.<strong title="Boolean">paused</strong>()</code></dt>
  <dd>
    Returns <code>true</code> if player is currently paused. Otherwise, returns
    <code>false</code>.
  </dd>

  <dt><code>mumbl.<strong title="Boolean">playing</strong>()</code></dt>
  <dd>
    Returns <code>true</code> if player is currently playing. Otherwise, returns
    <code>false</code>.
  </dd>
  
  <dt><code>mumbl.<strong title="Boolean">playerIs</strong>(<strong title="String">playerName</strong>)</code></dt>
  <dd>
    Returns <code>true</code> if the player being used is the player of
    <code>playerName</code>. Otherwise, returns <code>false</code>.
  </dd>

  <dt><code>mumbl.players.<strong title="Number (non-negative integer)">addPlayer</strong>(<strong title="String">playerName</strong>)</code></dt>
  <dd>
    Returns the player ID of the player of `playerName` (case-insensitive)</code>.
    If no such player already exists, it is added to <code>mumbl.players</code> and the
    player ID of the newly added player is returned.
  </dd>

  <dt><code>mumbl.<strong title="Number (non-negative)">position</strong>()</code></dt>
  <dd>
    Returns the current seek position, in seconds, of the currently selected track.
  </dd>

  <dt><code>mumbl.<strong title="void">position</strong>(<strong title="Number (non-negative)">newPosition</strong>)</code></dt>
  <dd>
    The currently selected track will seek to <code>newPosition</code>, in seconds.
  </dd>

  <dt><code>mumbl.<strong title="void">previous</strong>()</code></dt>
  <dd>Selects the previous track in the playlist from the currently selected track.</dd>

  <dt><code>mumbl.<strong title="Boolean">shuffle</strong>()</code></dt>
  <dd>
    Returns the shuffle state of the playlist.
  </dd>

  <dt><code>mumbl.<strong title="void">shuffle</strong>(<strong title="Boolean">shuffleState</strong>)</code></dt>
  <dd>
    If <code>shuffleState</code> is <code>true</code>, the playlist is played in a random
    order.
    <br />
    If <code>shuffleState</code> is <code>false</code>, the playlist is played in the
    order that it was created.
  </dd>


changelog.md
mumbl Changelog
===============

0.1.1b2
-------

* Fixed [issue #2](http://github.com/eligrey/mumbl/issues#issue/2 "mumbl.tracks() returns empty arrays when using HTML5").


0.1.1b1
-------

* Added better support for assigning media position in Songbird builds that have [bug
  21122](http://bugzilla.songbirdnest.com/show_bug.cgi?id=21122) fixed.
* Added support for getting the duration of the currently-playing media through
  `songbird.duration` in Songbird builds that have
  [bug 17578](http://bugzilla.songbirdnest.com/show_bug.cgi?id=17578) fixed.


0.1
---

* `addListener` has been renamed `observe` and `removeListener` has been renamed to
  `unobserve`.
* Errors from observers are no longer suppressed.
* `mumbl.onready` now passes `mumbl` as the first argument to any subscribed functions.
* Listeners added with `onready` are now called with `mumbl` as `this` by default.
* `INTEGRATED` renamed to `integrated`.
* `_interface` renamed to `interface`.
* Exposed `createEventDispatcher` to allow integrating events in third-party players.
* Revamped documentation with types as element titles instead of literally in the code
   among other improvements.
* Complete JSLint validation.
* Updated included SoundManager 2 version to 2.95b.20100323. This version or higher is
   now required.
* `mumbl.players.HTML5_AUDIO` renamed to `mumbl.players.HTML5`.
* Now using the new `soundManager.canPlayMIME` (the included SoundManager 2 has a
   has a slightly modified `soundManager.canPlayMIME` to fix a bug).
* No longer using the `_iO` property of sounds to change options as it may be changed
   in a future version of SoundManager 2.
* No longer using `Array.unshift`, which can be significantly slower than other methods.
* Now inspecting `arguments.length` instead of the type of the first argument.
* Removed support for Songbird versions below 1.4, as
   [bug 17577](http://bugzilla.songbirdnest.com/show_bug.cgi?id=17577) has been fixed.
* Cleaned up this changelog's Markdown.

### Bugfixes

* **Songbird**, **SoundManager 2**: Fixed media type regular expressions matching
   `media/typeANYTHING HERE` and being too strict and not matching ` media/type`.


0.1b3
-----

* `version` is now an array instead of a string. The first three items are version
  number parts and the fourth item is a string specifying the status of the version.
  The statuses are **a**lpha and **b**eta. The fifth and last item is the version of the
  status. `version.toString()` returns a human-readable version still.
* Better absolute URI resolving for Songbird.
* Changed terminology for "event handlers". A function passed to a listener method is
  now called an observer.
* The test suite has been improved and also now has the looping test automated. The
  points of human interaction are now only the "do you hear music" and "do you not hear
  music" questions.
* Event observers are now passed the event type they are observing.
* Removed `trackready` event from documentation. You should be listening for `duration`
  instead.

### Bugfixes

* **HTML5**: Fixed `clearChildren` internal method used to remove the children of th
  audio element.
* I previously forgot to mention in the changelog and update the documentation

mumbl.js
/*
 * mumbl JavaScript Library v0.1.1b2
 * 2010-10-28
 * By Eli Grey, http://eligrey.com
 *
 * See README.md for help
 * Licensed under the X11/MIT License
 *   See LICENSE.md
 */

/*global self, setTimeout */

/*jslint white: true, undef: true, eqeqeq: true, bitwise: true, regexp: true, nomen: true,
  strict: true, newcap: true, immed: true, maxlen: 90 */

/* Relevant Player Bugs
 *
 * Songbird:
 *  - [FIXED] http://bugzilla.songbirdnest.com/show_bug.cgi?id=21122
 *      Assignments to songbird.position get reverted
 *  - [FIXED] http://bugzilla.songbirdnest.com/show_bug.cgi?id=17577
 *      songbird.play() crashes Songbird unless playing and paused
 *  - [FIXED] http://bugzilla.songbirdnest.com/show_bug.cgi?id=17578
 *      songbird.duration or songbird.length needed for currently playing song
 *  - [FIXED] http://bugzilla.songbirdnest.com/show_bug.cgi?id=21122
 *      Assignments to songbird.position get reverted
 *  - [FIXED] http://bugzilla.songbirdnest.com/show_bug.cgi?id=17582
 *      accessing dataremotes through remoteapi via observers throws exception
 *  - http://bugzilla.songbirdnest.com/show_bug.cgi?id=15169
 *      Ogg-in-Flac files do not show track duration time
 *
 * Chromium / Google Chrome:
 *  - [FIXED] http://code.google.com/p/chromium/issues/detail?id=16768
 *      layout test fails because "ended" event is not triggered when video plays to end
 */

/*! @source http://purl.eligrey.com/github/mumbl/blob/master/mumbl.js*/

"use strict";

self.mumbl = self.mumbl || (function (self) {
	var
	
	// Fields:
	$player            = "player",
	$players           = $player + "s",
	$interface         = "interface",
	
	// Methods:
	$play              = "play",
	$stop              = "stop",
	$pause             = "pause",
	$position          = "position",
	$volume            = "volume",
	$playing           = $play  + "ing",
	$stopped           = $stop  + "ped",
	$paused            = $pause + "d",
	$loop              = "loop",
	$duration          = "duration",
	$add_track         = "addTrack",
	$remove_track      = "removeTrack",
	$add_tracks        = $add_track + "s",
	$next              = "next",
	$previous          = "previous",
	$clear             = "clear",
	$track             = "track",
	$tracks            = $track + "s",
	$length            = "length",
	$toggle_pause      = "togglePause",
	$mute              = "mute",
	$shuffle           = "shuffle",
	
	players            = 4,
	
	events             = {},
	played             = [],
	
	doc                = self.document,
	audioElem          = doc && doc.createElement("audio"),
	
	math               = Math,
	ArrayCtr           = Array,
	
	trackIndex         = 0,
	looping            = 2, // 0 = no loop, 1 = loop track, 2 = loop playlist
	duration           = 0,
	volume             = 1,
	playlistLen        = 0,
	
	TRUE               = !0,
	FALSE              = !1,
	NULL               = null,
	
	stopped            = TRUE,
	paused             = TRUE,
	muted              = FALSE,
	shuffled           = FALSE,
	
	$add_evt_listener  = "addEventListener",
	
	playlist,
	player,
	
	arrayIndexOf = ArrayCtr.prototype.indexOf || function (item) {
		var len = this[$length],
			i   = 0;
		for (; i < len; i++) {
			if (i in this && this[i] === item) {
				return i;
			}
		}
		return -1;
	},
	
	hasOwnProp = Object.prototype.hasOwnProperty,
	
	mumbl  = {
		version: [0, 1, 1, "b", 2], // 0.1.1b2
		player: 0,
		players: {
			UNSUPPORTED   : 0,
			HTML5         : 1,
			SONGBIRD      : 2,
			SOUNDMANAGER2 : 3,
			addPlayer     : function (player) { // use this to add custom players
				player = player.toUpperCase().replace(/\s/g, "");
				if (!(player in mumbl[$players])) {
					mumbl[$players][player] = players++;
				}
				return mumbl[$players][player];
			}
		},
		playerIs: function (player) {
			return mumbl.player === mumbl[$players][
				player.toUpperCase().replace(/\s/g, "")
			];
		},
		observe: function (event, observer) {
			if (hasOwnProp.call(events, event)) {
				events[event].push(observer);
			}
		},
		unobserve: function (eventType, observer) {
			if (hasOwnProp.call(events, eventType)) {
				var event = events[eventType],
				indexOfObserver = arrayIndexOf.call(event, observer);
				if (indexOfObserver !== -1) {
					event.splice(indexOfObserver, 1);
				}
			}
		},
		destruct: function () {
			mumbl[$clear]();
			delete self.mumbl;
		},
		onready: function (callback, scope) {
			scope = scope === NULL || typeof scope === "undefined" ? mumbl : scope;
			if (mumbl.player !== 3) { // !mumbl.playerIs("SoundManager")
				callback.call(scope, mumbl);
			} else {
				player.onready(function () {
					callback.call(scope, mumbl);
				});
			}
		},
		integrated: FALSE
	},
	
	removeFromPlayed = function (track) {
		var i = played[$length];
		while (i--) {
			if (played[i] > track) {
				played[i]--;
			} else if (played[i] === track) {
				played.splice(i, 1);
			}
		}
	},
	
	randomUnplayedTrack = function () {
		var track = math.floor(math.random() * playlistLen);
		if (arrayIndexOf.call(played, track) === -1) {
			return track;
		}
		return randomUnplayedTrack();
	},
	
	onEnded = function () {
		var track = mumbl[$track]();
		if (looping) {
			if (looping === 2 && track === mumbl[$length]() - 1) {
				mumbl[$track](0);
			} else if (looping === 1) {
				mumbl[$track](track);
				return;
			}
		}
		if (track < mumbl[$length]() - 1) {
			mumbl[$next]();
		}
	},
	
	slice = function (obj) {
		return ArrayCtr.prototype.slice.call(obj);
	},
	
	createExceptionThrower = function (ex) {
		return function () {
			throw ex;
		};
	},
	
	dispatchEvent = function (eventType) {
		var event = events[eventType],
		len = event[$length],
		i = 0;
		if (len) {
			for (; i < len; i++) {
				try {
					event[i].call(mumbl, eventType);
				} catch (ex) {
					setTimeout(createExceptionThrower(ex), 0);
				}
			}
		}
	},
	
	createEventDispatcher = mumbl.createEventDispatcher = function (event) {
		if (!hasOwnProp.call(events, event)) {
			events[event] = [];
		}
		return function () {
			dispatchEvent(event);
		};
	},
	
	// Event methods:
	
	$canplay  = "canplay",
	$external = "external",
	
	onTrackChange             = createEventDispatcher($track),
	onDurationUpdate          = createEventDispatcher($duration),
	onPositionUpdate          = createEventDispatcher($position),
	onCanPlayTrack            = createEventDispatcher($canplay),
	onCanPlayThroughTrack     = createEventDispatcher($canplay + "through"),
	onTrackLoad               = createEventDispatcher("load"),
	onExternalPlayStateChange = createEventDispatcher($external + "playstate"),
	onExternalVolumeChange    = createEventDispatcher($external + "volume"),
	onExternalMuteChange      = createEventDispatcher($external + "mute"),
	onExternalLoopingChange   = createEventDispatcher($external + "looping"),
	onExternalShufflingChange = createEventDispatcher($external + "shuffling"),
	
	allTrackHandlers = function () {
		onTrackChange();
		onDurationUpdate();
		onPositionUpdate();
		onCanPlayTrack();
		onCanPlayThroughTrack();
		onTrackLoad();
	},
	
	inited;
	
	mumbl.version.toString = function () {
		return this.slice(0, 3).join(".") + this.slice(3, 5).join("");
	};
	
	mumbl[$length] = function () {
		return playlistLen;
	};
	mumbl[$playing] = function () {
		return !stopped;
	};
	mumbl[$paused] = function () {
		return paused;
	};
	mumbl[$stopped] = function () {
		return stopped;
	};
	mumbl[$next] = function () {
		var item = mumbl[$track];
		if (!shuffled) {
			if (item() + 1 < mumbl[$length]()) {
				item(item() + 1);
			} else if (looping === 2) {
				item(0);
			}
		} else {
			if (played[$length] === playlistLen) { // played every track
				if (looping === 2) {
					played[$length] = 0; // reset played
				} else {
					return;
				}
			}
			
			var track = arrayIndexOf.call(played, item()) + 1 && played.push(item());
			if (played[track]) {
				item(played[track]);
			} else {
				if (played[$length] === playlistLen) {
					played[$length] = 0;
				}
				item(randomUnplayedTrack());
			}
		}
	};
	mumbl[$previous] = function () {
		var item = mumbl[$track];
		if (!shuffled) {
			if (item()) {
				item(item() - 1);
			} else if (looping === 2) {
				item(playlistLen - 1);
			}
		} else {
			var track = arrayIndexOf.call(played, item());
			
			if (track === -1) {
				track = played.push(item()) - 1;
			}
			if (track) {
				item(played[track - 1]);
			} else {
				mumbl[$next]();
			}
		}
	};
	mumbl[$toggle_pause] = function () {
		if (paused) {
			mumbl[$play]();
		} else {
			mumbl[$pause]();
		}
	};
	mumbl[$shuffle] = function (state) {
		if (!arguments[$length]) {
			return shuffled;
		}
		shuffled = state;
	};
	mumbl[$add_tracks] = function () {
		var args = slice(arguments),
		len = args[$length],
		i = 0;
		for (; i < len; i++) {
			mumbl[$add_track].apply(mumbl, args[i]);
		}
	};
	mumbl[$duration] = function () {
		return duration;
	};
	
	inited = (self.songbird && (function () {
		player               = self.songbird;
		
		var currentTrack     = 0,
		notSettingVolume     = TRUE,
		notChangingPlayState = TRUE,
		notMuting            = TRUE,
		notSettingShuffle    = TRUE,
		notSettingLooping    = TRUE,
		canSetPosition       = TRUE,
		library              = player.siteLibrary,
		durationProp         = "http://songbirdnest.com/data/1.0#duration",
		uriProp              = "http://purl.eligrey.com/mumbl#originURI",
		faceplateEventType   = "faceplate.",
		anchor               = doc.createElement("a"),
		isOGG                = /^\s*a(?:udio|pplication)\/(?:x-)?ogg\s*(?:$|;)/i,
		addListener          = function (topic, observer) {
			player.addListener(topic, {
				observe: observer
			});
		},
		playlist = library.createSimpleMediaList("mumbl"),
		tempPos;
		playlist[$clear](); // just in case it wasn't cleared previously (browser crash)
		
		try { // This throws an error pre-bug 21122 fix
			tempPos = player[$position];
			player[$position] = 0;
			player[$position] = tempPos;
			tempPos = NULL;
		} catch (e) {
			canSetPosition = FALSE;
		}
		
		mumbl[$player]    = 2; // mumbl.players.SONGBIRD
		mumbl[$interface] = player;
		mumbl.integrated  = TRUE;
		
		mumbl[$play] = function () {
			notChangingPlayState = FALSE;
			player[$play]();
			notChangingPlayState = TRUE;
			stopped = FALSE;
		};
		mumbl[$pause] = function () {
			notChangingPlayState = FALSE;
			player[$pause]();
			notChangingPlayState = TRUE;
		};
		mumbl[$add_track] = function (uri) {
			// Songbird can play anything gstreamer can but has issues with seeking OGG
			// files, so prefer anything that doesn't match ^a(udio|pplication)\/ogg
			if (arguments[$length] > 2) { // more than one file specified
				var args = slice(arguments),
				len = args[$length],
				file,
				i = 0;
				
				for (; i < len; i += 2) {
					file = args.slice(i, i + 2);
					// file[0] = file URI
					// file[1] = file media type
					if (!isOGG.test(file[1])) {
						uri = file[0];
						break;
					}
				}
			}
			anchor.href = uri;
			var absURI = anchor.href, // get absolute location of uri
			mediaItem  = library.createMediaItem(absURI);
			anchor.removeAttribute("href");
			mediaItem.setProperty(uriProp, uri);
			playlist.add(mediaItem);
			
			playlistLen = playlist[$length];
		};
		mumbl[$stop] = function () {
			player[$stop]();
			duration = 0;
			stopped = TRUE;
			allTrackHandlers();
		};
		mumbl[$mute] = function (state) {
			if (!arguments[$length]) {
				return muted;
			}
			notMuting = FALSE;
			player[$mute] = muted = state;
			notMuting = TRUE;
		};
		mumbl[$loop] = function (loopType) {
			if (!arguments[$length]) {
				return looping;
			}
			notSettingLooping = FALSE;
			player.repeat = loopType;
			notSettingLooping = TRUE;
		};
		mumbl[$shuffle] = function (state) {
			if (!arguments[$length]) {
				return shuffled;
			}
			notSettingShuffle = FALSE;
			player[$shuffle] = state;
			notSettingShuffle = TRUE;
		};
		mumbl[$tracks] = function () {
			// recreate playlist as a TrackItemList
			var tracks = [],
			    i      = 0;
			for (; i < playlistLen; i++) {
				tracks.push([playlist.getItemByIndex(i).getProperty(uriProp), NULL]);
			}
			return tracks;
		};
		mumbl[$volume] = function (newVolume) {
			if (!arguments[$length]) {
				return volume;
			}
			notSettingVolume = FALSE;
			player[$volume] = newVolume * 255;
			notSettingVolume = TRUE;
		};
		mumbl[$track] = function (index) {
			if (!arguments[$length]) {
				return trackIndex;
			}
			var wasPaused = paused;
			notChangingPlayState = FALSE;
			player.playMediaList(playlist, index);
			if (wasPaused) {
				mumbl[$pause]();
			}
			notChangingPlayState = TRUE;
		};
		mumbl[$remove_track] = function (index) {
			playlist.removeByIndex(index);
			playlistLen = playlist[$length];
		};
		mumbl[$clear] = function () {
			mumbl[$stop]();
			playlist[$clear]();
			playlistLen = playlist[$length];
		};
		mumbl[$playing] = function () {
			return player[$playing];
		};
		mumbl[$position] = function (seconds) {
			if (!arguments[$length]) {
				return player[$position] / 1000;
			} else if (canSetPosition && !mumbl[$stopped]()) {
				if (seconds < 0) {
					seconds = 0;
				}
				// Only works in builds with Songbird bug 21122 fixed
				player[$position] = seconds * 1000; // convert seconds to ms
			}
		};
		mumbl.destruct = function () {
			mumbl[$clear]();
			library.remove(playlist);
			delete self.mumbl;
		};
		
		addListener(faceplateEventType + $volume, function () {
			volume = player[$volume] / 255;
			if (notSettingVolume) {
				onExternalVolumeChange();
			}
		});
		addListener(faceplateEventType + $mute, function () {
			muted = player[$mute];
			if (notMuting) {
				onExternalMuteChange();
			}
		});
		addListener(faceplateEventType + $paused, function () {
			paused = player[$paused];
			if (notChangingPlayState) {
				onExternalPlayStateChange();
			}
		});
		addListener("playlist.repeat", function () {
			looping = player.repeat;
			if (notSettingLooping) {
				onExternalLoopingChange();
			}
		});
		addListener("playlist." + $shuffle, function () {
			shuffled = player[$shuffle];
			if (notSettingShuffle) {
				onExternalShufflingChange();
			}
		});
		addListener("metadata." + $position, onPositionUpdate);
		
		self[$add_evt_listener]("trackchange", function (evt) {
			var track = evt.item;
			if (!playlist.contains(track)) {
				return;
			}
			currentTrack = track;
			trackIndex   = playlist.indexOf(track);
			
			allTrackHandlers();
		}, FALSE);
		
		if ($duration in player) { // Songbird bug 17578 fixed
			mumbl[$duration] = function () {
				return player[$duration] * 1000; // convert ms to seconds
			};	
		} else {
			self[$add_evt_listener]("trackchange", function (evt) {
				duration     = (
					parseInt(
						evt.item.getProperty(durationProp),
						10
					) / 1000000 // convert microseconds to seconds
				) || 0;
				onDurationUpdate();
			}, FALSE);
		}
		
		self[$add_evt_listener]("unload", mumbl.destruct, FALSE);
		
		return (paused = (stopped = TRUE));
	}())) || (typeof audioElem.canPlayType === "function" && (function () {
		player   = audioElem;
		playlist = [];
		
		var addEvent,
		clearChildren = function () {
			var nodes = player.childNodes,
			i = nodes[$length];
			while (i--) {
				player.removeChild(nodes.item(i));
			}
			player.load();
		};
		
		if (player[$add_evt_listener]) {
			addEvent = function (event, observer) {
				player[$add_evt_listener](event, observer, FALSE);
			};
		} else if (player.attachEvent) {
			// you never know, IE might actually add support for HTML5 audio
			addEvent = function (event, observer) {
				player.attachEvent("on" + event, observer);
			};
		}
	
		mumbl[$player]    = 1; // mumbl.players.HTML5_AUDIO
		mumbl[$interface] = player;
		
		
		mumbl[$play] = function () {
			player[$play]();
			(paused = stopped = FALSE);
		};
		mumbl[$pause] = function () {
			player[$pause]();
			paused = !(stopped = FALSE);
		};
		mumbl[$stop] = function () {
			stopped = TRUE;
			duration = 0;
			clearChildren();
			allTrackHandlers();
		};		
		mumbl[$clear] = function () {
			mumbl[$stop]();
			playlist = [];
			played   = [];
			playlistLen = 0;
		};
		mumbl[$add_track] = function () {
			var args  = slice(arguments),
			    len   = args[$length],
			    files = [],
			    i     = 0;
			for (; i < len; i += 2) {
				files.push(args.slice(i, i + 2));
			}
			playlistLen = playlist.push(files);
		};
		mumbl[$remove_track] = function (index) {
			playlist.splice(index, 1);
			playlistLen = playlist[$length];
			removeFromPlayed(index);
		};
		mumbl[$tracks] = function () {
			// convert [[[1,2],[3,4]],[[5,6],[7,8]]] to [[1,2,3,4],[5,6,7,8]]
			var flatItem = [],
			flatPlaylist = [],
			i            = 0,
			items, item;
			for (; i < playlistLen; i++) {
				items = playlist[i];
				item = items[$length];
				while (item--) {
					flatItem = items[item].concat(flatItem);
				}
				flatPlaylist.push(flatItem.slice(0));
				flatItem.length = 0;
			}
			return flatPlaylist;
		};
		mumbl[$loop] = function (state) {
			if (!arguments[$length]) {
				return looping;
			}
			if ((looping = state) && player.ended) {
				onEnded();
			}
		};
		mumbl[$mute] = function (state) {
			if (!arguments[$length]) {
				return player.muted;
			}
			muted = player.muted = state;
		};
		mumbl[$track] = function (index) {
			if (!arguments[$length]) {
				return trackIndex;
			}
			var items = playlist[index],
			len = items[$length],
			source,
			i = 0;
			clearChildren();
			for (; i < len; i++) {
				source = doc.createElement("source");
				source.setAttribute("src", items[i][0]);
				if (items[i][1]) {
					source.setAttribute("type", items[i][1]);
				}
				player.appendChild(source);
			}
			trackIndex = index;
			stopped = FALSE;
			player.load();
			if (paused) {
				player[$pause]();
			}
			onTrackChange();
		};
		mumbl[$position] = function (seconds) {
			if (!arguments[$length]) {
				return player.currentTime;
			} else if (!stopped) {
				if (seconds < 1) {
					// fix negative position
					seconds = (seconds < 0 ? 0 : seconds) + 0.01022;
				}
				player.currentTime = seconds;
			}
		};
		mumbl[$volume] = function (vol) {
			if (!arguments[$length]) {
				return volume;
			}
			volume = player[$volume] = vol;
		};
		
		addEvent($duration + "change", function () {
			duration = player[$duration];
			onDurationUpdate();
		});
		addEvent("timeupdate", onPositionUpdate);
		addEvent("suspend", onTrackLoad); // most likely loaded
		addEvent("load", onTrackLoad);
		addEvent("canplay", onCanPlayTrack);
		addEvent("canplaythrough", onCanPlayThroughTrack);
		addEvent("ended", onEnded);
		
		return (player.autoplay = TRUE);
	}())) || (self.soundManager && (function () {
		// http://www.schillmania.com/projects/soundmanager2/doc/
		player   = self.soundManager;
		playlist = [];
		
		player.useFastPolling = player.allowPolling = TRUE;
		
		var sounds     = player.sounds,
		mumblNS        = "mumbl:",
		currentTrackId,
		getCurrentTrack = function () {
			return sounds[currentTrackId];
		},
		playSound = function (sound) {
			sound.setVolume(volume * 100);
			sound.setPosition(0);
			sound[$play]();
			if (sound.loaded) {
				allTrackHandlers();
			}
		},
		createReadyHandler = function (trackId) {
			var track = sounds[trackId],
			    ready = FALSE;
			return function () {
				if (ready) {
					onDurationUpdate();
				} else if (track[$duration] > track[$position] - 1) {
					// > 1 second is loaded
					ready = TRUE;
					onDurationUpdate();
					onCanPlayTrack();
				}
			};
		},
		createPlayingHandler = function (trackId) {
			var track = sounds[trackId],
			    ready = FALSE;
			return function () {
				if (ready) {
					onPositionUpdate();
				} else {
					ready = TRUE;
					if (stopped || paused) {
						track[$pause]();
						track.setPosition(0);
					}
					onDurationUpdate();
					onCanPlayThroughTrack();
				}
			};
		},
		SM2SoundOpts = function (uri) {
			var opts = this,
			trackId  = opts.id = mumblNS + (opts.url = uri);
			
			// SM2 disregards inherited properties, so don't use SoundOpts.prototype
			
			// callbacks
			opts.onload                 = allTrackHandlers;
			opts.whileloading           = createReadyHandler(trackId);
			opts.whileplaying           = createPlayingHandler(trackId);
			opts.onfinish               = onEnded;
			
			opts.stream                 = TRUE;
			
			// null
			opts.onplay                 = 
			opts.onpause                = 
			opts.onresume               = 
			opts.onstop                 = 
			opts.onid3                  = 
			opts.onbeforefinish         = 
			opts.onbeforefinishcomplete = 
			opts.onjustbeforefinish     = 
			opts.position               = NULL;
			
			// false
			opts.autoLoad               = 
			opts.multiShot              = 
			opts.multiShotEvents        = 
			opts.autoPlay               = FALSE;
			
			// ints
			opts.onbeforefinishtime     = // default 5000
			opts.onjustbeforefinishtime = // default 200
			opts.pan                    = 0;
			opts.volume                 = 100;
		};
		
		mumbl[$player]    = 3; // mumbl.players.SOUNDMANAGER2
		mumbl[$interface] = player;
		
		mumbl[$play] = function () {
			var track = getCurrentTrack();
			if (track && (stopped || paused)) {
				track.resume();
				stopped = paused = FALSE;
			}
		};
		mumbl[$pause] = function () {
			getCurrentTrack()[$pause]();
			paused = TRUE;
		};
		mumbl[$stop] = function () {
			getCurrentTrack()[$stop]();
			paused = stopped = TRUE;
		};
		mumbl[$loop] = function (state) {
			if (!arguments[$length]) {
				return looping;
			}
			if ((looping = state) &&
			    math.floor(mumbl[$position]() / mumbl[$duration]()) === 1)
			{
				onEnded();
			}
		};
		mumbl[$position] = function (seconds) {
			if (!arguments[$length]) {
				return getCurrentTrack()[$position] / 1000 || 0;
			}
			getCurrentTrack().setPosition(seconds * 1000);
		};
		mumbl[$volume] = function (vol) {
			if (!arguments[$length]) {
				return volume;
			}
			volume = vol;
			if (currentTrackId) {
				getCurrentTrack().setVolume(vol * 100);
			}
		};
		mumbl[$track] = function (index) {
			if (!arguments[$length]) {
				return trackIndex;
			}
			if (playlistLen) {
				var sound = sounds[playlist[index]];
				if (sound) {
					if (currentTrackId) {
						getCurrentTrack()[$stop]();
					}
					playSound(sound);
					currentTrackId = playlist[trackIndex = index];
					onTrackChange();
				}
			}
		};
		mumbl[$tracks] = function () {
			var tracks = [],
			    i      = 0;
			for (; i < playlistLen; i++) {
				tracks.push([playlist[i].substr(6), NULL]); // 6 === mumblNS.length
			}
			return tracks;
		};
		mumbl[$mute] = function (state) {
			if (!arguments[$length]) {
				return getCurrentTrack().muted;
			}
			getCurrentTrack()[(state ? "" : "un") + $mute]();
		};
		mumbl[$duration] = function () {
			return (getCurrentTrack().durationEstimate / 1000) || 0;
		};
		mumbl[$add_track] = function (uri) {
			if (arguments[$length] > 2) { // more than one file specified
				var args = slice(arguments),
				len = args[$length],
				file,
				i = 0;
				for (; i < len; i += 2) {
					file = args.slice(i, i + 2);
					// file[0] = file URI
					// file[1] = file media type
					if (player.canPlayMIME(file[1])) {
						uri = file[0];
						break;
					}
				}
			}
			player.createSound(new SM2SoundOpts(uri));
			playlistLen = playlist.push(mumblNS + uri);
		};
		mumbl[$remove_track] = function (index) {
			sounds[playlist[index]].destruct();
			playlist.splice(index, 1);
			playlistLen = playlist[$length];
			removeFromPlayed(index);
		};
		mumbl[$clear] = function () {
			var i = playlistLen;
			while (i--) {
				mumbl[$remove_track](i);
			}
			currentTrackId = FALSE;
		};
	}()));
	
	return mumbl;
}(self));
